packages/@aws-cdk/aws-codebuild/test/test.project.ts

```diff
@@ -1466,12 +1475,236 @@ export = {
+                  ':secret:secretId-??????',
+                ]],
+              },
+            }),
+          },
+        }));
+
+        test.done();
+      },
+
+      'can be provided as a SecretValue with a jsonField and versionId'(test: Test) {
+        // GIVEN
+        const stack = new cdk.Stack();
+        // WHEN
+        const secretValue = cdk.SecretValue.secretsManager(
+          'arn:aws:secretsmanager:us-west-2:123456789012:secret:my-secret-123456',
+          { jsonField: 'json-key', versionId: 'version-id' },
+        );
+        new codebuild.PipelineProject(stack, 'Project', {
+          environmentVariables: {
+            'ENV_VAR1': {
+              type: codebuild.BuildEnvironmentVariableType.SECRETS_MANAGER,
+              value: secretValue,
+            },
+          },
+        });
+
+        // THEN
+        expect(stack).to(haveResourceLike('AWS::CodeBuild::Project', {
+          'Environment': {
+            'EnvironmentVariables': [
+              {
+                'Name': 'ENV_VAR1',
+                'Type': 'SECRETS_MANAGER',
+                'Value': 'arn:aws:secretsmanager:us-west-2:123456789012:secret:my-secret-123456:json-key:version-id',
+              },
+            ],
+          },
+        }));
+
+        // THEN
+        expect(stack).to(haveResourceLike('AWS::IAM::Policy', {
+          'PolicyDocument': {
+            'Statement': arrayWith({
+              'Action': 'secretsmanager:GetSecretValue',
+              'Effect': 'Allow',
+              'Resource': 'arn:aws:secretsmanager:us-west-2:123456789012:secret:my-secret-123456*',
+            }),
+          },
+        }));
+
+        test.done();
+      },
+
+      'throws error when SecretValue is provided with versionStage and versionId'(test: Test) {
+        // GIVEN
+        const stack = new cdk.Stack();
+
+        // WHEN
+        const secretValue = cdk.SecretValue.secretsManager('secretId', { versionStage: 'version-stage', versionId: 'version-id' });
+
+        // THEN
+        test.throws(() => {
+          new codebuild.PipelineProject(stack, 'Project', {
+            environmentVariables: {
+              'ENV_VAR1': {
+                type: codebuild.BuildEnvironmentVariableType.SECRETS_MANAGER,
+                value: secretValue,
+              },
+            },
+          });
+        }, /Secret Value cannot provide versionStage and versionId at the same time!/);
+
+        test.done();
+      },
     },
 
     'should fail creating when using a secret value in a plaintext variable'(test: Test) {
       // GIVEN
       const stack = new cdk.Stack();
 
+      // THEN
+      test.throws(() => {
+        new codebuild.PipelineProject(stack, 'Project', {
+          environmentVariables: {
+            'a': {
+              value: cdk.SecretValue.secretsManager('my-secret'),
+            },
+          },
+        });
+      }, /Plaintext environment variable 'a' contains a secret value!/);
+
+      test.done();
+    },
```

Comment:
Why do we need this test? It's not really related to this PR, right?

Reply:
This test came to my mind when I saw that a test in `aws-codebuild-actions` was failing but not in `was-codebuild`.
https://github.com/aws/aws-cdk/blob/6ec9aa1ec7fc21b425608c14880d0d712a3aa79a/packages/%40aws-cdk/aws-codebuild/test/test.project.ts#L1704-L1720
This test only validates that when we pass a `SecretValue` as a token and we use `PLAINTEXT` as type that it throws an error.
Passing it as `SecretValue` direct did not throw an error and thus I added this test. And this implementation to fix it. https://github.com/aws/aws-cdk/blob/6ec9aa1ec7fc21b425608c14880d0d712a3aa79a/packages/%40aws-cdk/aws-codebuild/lib/project.ts#L729-L736

Not the brightest way to fix it xD now your additional if statement makes this test pass.